### ISSUE TITLE
Universal IDs for widgets and other objects

### ISSUE THREAD

**Issue by johnspackman** [qooxdoo/qooxdoo#9422]
It is useful to be able to identify widgets by an ID, whether for UI testing via a tool like Selenium or for general coding organisation.  An example of Selenium testing (taken from a traditional web app test) is:

```
describe("Contact form", () => {
  test("lead can submit a contact request", async () => {
    await page.waitForSelector("[data-test=contact-form]");
    await page.click("input[name=name]");
    await page.type("input[name=name]", lead.name);
    await page.click("button[type=submit]");
    await page.waitForSelector(".modal");
  }, 16000);
});
```

Another use for IDs is during development as form of centralising widget creation in a form that can be overridden by derived classes - `qx.ui.core.Widget` has `getChildControl` and `_createChildControlImpl` which are specifically targeted at the construction of UI components, and the IDs are used for selecting appearance.  However, this is targeted at the appearance mechanism and is very low level; it is not suitable for application level identification (and attempts to do so will break because of what the appearance mechanism expects to find there).

## Parent/Child or Owner/Objects
There is benefit in having an ID mechanism for objects which is more application-centric; similar to `getChildControl`/`_createChildControlImpl` this would allow construction of objects to be centralised and allow derived classes to override features.  Note that this should support any `qx.core.Object` and not just widgets.

At an application level, a particular form would have various fields (or other objects) that need to be identified, but the structure does not necessarily bear any relation to the widget `layoutParent`/`children` relationship.  For example, consider a UserProfileEditor which has TextField's for `personName` and `emailAddress`, and a SelectBox for `company`; whether you place them all in one Composite with a Grid layout, or use two or three Composites with VBox/HBox layouts has nothing to do with the application-level structuring of the fields you want to identify.

To avoid confusion, we could call this application-level relationship Owner/Objects, where we add `getObject` and `_createObjectImpl` methods to some class in Qooxdoo (maybe `qx.core.Object` or as a mixin?)

## Universal identification
By enforcing basic rules on names (eg alphanumeric plus "-" and "_"), we can assemble ID paths like "contactForm/edtName" and navigate the hierarchy from a given start point.

This would imply that every instance of a widget has an absolute path based identifier, which would be great for debugging or anything machine based.

How this would work with Selenium I'm not sure - AIUI Selenium would require a Sizzle expression to select widgets, which is fine for simple applications but would presumably become exponentially harder where there are dozens (or more) of forms in an application.  Is it possible to add hooks into Selenium where we can add our own ID navigation?

**Comment by cboulanger** [comment]
I don't know anything about Selenium, just [getting my hands dirty](https://github.com/cboulanger/qx-puppeteer) with [puppeteer](https://www.valentinog.com/blog/ui-testing-jest-puppetteer/) - from which the above example comes. I completely agree with your approach  - we should agree on a naming convention and support that within qooxdoo, and make it generic enough that different testing frameworks can interface with it.

One way would be to follow your Owner/Objects naming convention and then set a `data-*` attribute on the DOM node of the (parent/child) container widget. This way, it can be found by any testing framework.

**Comment by cboulanger** [comment]
I wonder what would be better - to have an arbitrary string as ID, which would, on the basis of a convention, be parsed into components to make it meaningful, such as "ownerId/objectID", or to enforce a structure with an API such as `anyqxobject.set({ownerId:"ownerId", objectId:"objectId"})`.  The first option would be more flexible and would allow a 1:1 correspondence between data-* attributes and ID. Drawback would be that each target object would have a static ID. The second seems more portable, since it would allow parent-child relationships that could be moved between higher-level owner-objects.

**Comment by derrell** [comment]
Is `ownerId` actually the parent's `objectId`? If not, then it might be useful to use the latter (map) capability, and add a `parentId` feature that would allow following the path up to figure out which particular widget is involved. (This would be very useful when inspecting the DOM and trying to figure out exactly what/where you're looking at.) I do realize this is diverging from the original intent of this issue...

**Comment by cboulanger** [comment]
On further reflection, the second option seems more flexible and versatile. Since the GUI test framework could use selectors to identify the widget's DOM node (one can even [use XPATH with puppeteer](https://github.com/GoogleChrome/puppeteer/issues/537)), ownerID and objectID could be separate data-* attributes.

**Comment by cboulanger** [comment]
@derrell As I understand @johnspackman's proposal, `ownerId` is indeed the parent's `objectId`. qooxdoo would have to keep a registry which then can be queried to construct the parent/child path  - This cries for a component which can inspect these relationships with a Tree widget :-)

**Comment by cboulanger** [comment]
Here's a minimal implementation:
https://github.com/cboulanger/qx-puppeteer/blob/master/demo/source/class/demo/MObjectId.js
it works like this:
https://github.com/cboulanger/qx-puppeteer/blob/master/demo/source/class/demo/Application.js#L34
and this test:
https://github.com/cboulanger/qx-puppeteer/blob/master/tests/demo.spec.js#L46
@johnspackman Is this roughly what you had in mind?

**Comment by derrell** [comment]
Having an actual implementation is useful. Thanks! I see a few potential problems:

- It's a PITA to add these all the time. Maybe we just need to live with it, if we want them, but it sure would be nice to find a way to automate the process.
- More importantly, those IDs will get out of sync whenever you move code around. Automating the process, at least as far as setting the `parentId` (which I would, in fact, rename from `ownerId`). Whenever an object is added to a container, its `parentId` could be set automatically.
- It currently depends on an "appear" event, but this should work with any qx.core.Object, not just with widgets, and in fact, should work with non-visible widgets as well

I may be viewing this from a different vantage point than others?

**Comment by cboulanger** [comment]
@derrell thank you for your comments! All very good points that should be addressed... One thing: we should probably automate this only in debug/test mode, because otherwise we needlessly create DOM properties...

**Comment by cboulanger** [comment]
What about a default object ID which is the class name with an incrementing instance count (such as qx.ui.window.Window[1] (qx.ui.window.Window[2], ...), which would allow paths like qx.ui.window.Window[1]/qx.ui.form.Button[1]. 

Maybe we could turn automatic object id assignment on and off through a flag, the default being off.

**Comment by level420** [comment]
@cboulanger good idea! But it may be a better to not have dots and square brackets in the names as those tend to interfere with languages syntax like xpath etc. maybe just `qx_ui_form_Button_1` or similar.

**Comment by level420** [comment]
and it would be really helpful to have the enumeration relative to the parent to allow adressing the first child of a parent like `qx_ui_window_1/qx_ui_form_button_3` to find the third button instance in the first window instance.

**Comment by johnspackman** [comment]
An `ownerId` will not work because it's not scalable and because it forces us back into a situation where there is a list of all objects in a map somewhere (i.e. remove automatic memory management); it's not scalable because it would imply that every ID was globally unique.  My proposal is more that the Object would have a map of child object IDs to child objects, and each object would have an ID {String} and an Owner {qx.core.Object}.

The globally unique ID part would come because you could have an ID made up of paths, eg "/application/usersEditor/userProfileEditor/username" that could be navigated by a fairly simple iteration function.

I'm not convinced about relative IDs via a subscript (eg "Window[1]") either because it seems quite fragile - especially if it is arranged around UI components rather then logical/business components - what would happen if things were based around `Composite[2]`, but that was subsequently changed from a Composite with a Grid layout to 3 Composites with VBox/HBox layouts?  The unit tests should not have to change because of an implementation detail.

@derrell although I quite like the "parent/child" naming, that's kind of already taken with widgets and their children - although widget parents are called "layoutParent", children are "children" and so a noticably different naming could be really useful at times (I'm not tied to owner/owned, just that I think we should steer clear of "parent" and "child")

WRT how much of a PITA it is to assign individual IDs to widgets, I think it becomes more interesting if IDs are something that is assigned naturally - for example, if objects are created via calls to `this.getObject("edtUsername")` then they implicitly have an ID; this is the same pattern as the `myWidget.getChildControl("icon")`.  From an application point of view, the reason you might start to use this pattern is that (a) it removes the spaghetti of ordering and member variables during form construction, and (b) it allows derived classes to replace objects in a controller manner.

**Comment by johnspackman** [comment]
PS my proposal does not cater well for existing applications, which I think adds a significant complication - the issue is that assigning an ID to a widget is easy, but making it possible to locate that widget via a globally unique equivalent is hard.  In my proposal, IDs are unique because they are relative to their parent/owner, and this implies that the "owner" hierarchy is well defined.  Existing applications will not have a regular "owner" hierarchy, or a means of navigating widgets in an application-centric manner, hence the need to use something like `Window[2]` which is so fragile.

**Comment by cboulanger** [comment]
Can't we take up the "control" vocabulary - so inside widgets, we have "child controls", inside an application we have controls and "subcontrols". This would allow us to create control hierarchies that reach down to the child controls, such as "/editor/sendButton/label". Assigning the data-* attributes for child controls can be done automatically since they already have an ID.

**Comment by cboulanger** [comment]
I just wonder how, without a central registry, you can find, given a path, the top "controls/parents" which allow you to drill into their "subcontrol/childrens"

**Comment by johnspackman** [comment]
"Control" implies a UI control so it isn't ideal because it's often more than controls, eg the pattern for creating objects by ID is useful for creating `qx.data.controller.*` and other common objects.

To locate something without a registry, you just have to have a starting point; that could be some global object which registers `qx.core.Init.getApplication()` as (eg) "application" so that paths would be "/application/this/that/usersEditor/userEditor/name".

Relative paths would be the same so something like:
```
var usersEditor = qx.core.Ids.getObject("/application/this/that/usersEditor");
var txtUserName = usersEditor.getObject("userEditor/name");
```

**Comment by cboulanger** [comment]
Sounds great to me. Very simple API. I have a bias towards UI controls because I am interested in GUI tests. But you are totally right that this needs to be a more generic system. 

One more thought:  I wonder if we should design the system so that it could store some meta-data about the structure of the application that could be visualized using the API viewer, for example, an extra tab on the left "app structure" and that could be exported to allow third parties (such as plugin developers) to hook into the app. This is a problem I have been trying to solve with my app forever, using all kind of hacks, and would love to see as an official qooxdoo feature.

**Comment by johnspackman** [comment]
I wrote up a quick gist of what I'm imagining, based on your MObjectId but not tested, just written up: https://gist.github.com/johnspackman/f109bae1756e6f19f30c429c007f9c95

Re application structure: yes, that could be really really cool and I think that this ID mechanism would enable it; there was/is an inspector somewhere that you could add in to browse your app (at the Widget level).  AFAICR it was a bit too low level and not straightforward to get hooked in so I never used it.  There are so many times when, for example, you want to know what is in the model of the `qx.data.controller.List` that populates a certain SelectBox but you're not in the debugger to be able to get a pointer to it.

**Comment by cboulanger** [comment]
Simple and elegant. I am in favour of this approach and would support a PR in this direction. The meta-data/documentation I was thinking of could be stored in the comments, I guess that the ID is all we need in the code itself. I am thinking of some field that would map ids to "descriptions" that explain what the object does in terms of application/business logic (not as a **class**, but as a **feature**).

**Comment by cboulanger** [comment]
Maybe the automated id assignment that @cajus and @derrell would like to see could be implemented by including separate "homemade" mixins that would allow to use their already established patterns? I think the mechanism we provide on the framework level should be as open and generic as possible.

**Comment by johnspackman** [comment]
ah, ISWYM - a great way to do that would be to use annotations because it is an existing mechanism that would save us from having to define what meta data could to be available.  The ID mechanism is simply a way to navigate the structure, so if you added an annotation specific to libraries or researchers in bibliograph it would be up to some other tool (eg an "Inspector" widget) to decide how to interpret that meta data.  Similarly, adding a generic "description" annotation is just about defining that annotation and having classes reference it.

**Comment by johnspackman** [comment]
WRT automating the process, once IDs are assigned I think most of the automation is already taken care of - I updated that gist just now to fix the missing code for this, but when you call `this.getObject("txtName")` and it calls `this._createObjectImpl`, the returned object has `obj.setOwner(this)` passed to it.

Also,  by calling `widget.getContentElement().setAttribute(...)` the IDs are automatically added and removed as the widget gets bound to the DOM.

What is left is assigning a globally unique ID to each DOM element, which could be expensive unless it is carefully cached - i.e. translating the username text field's ID of `"name"` to `"_application_this_that_usersEditor_userEditor_name"`.  I imagine that this will be pretty essential to use with selenium, unless there is a way of hooking into Selenium's selection mechanism and adding a plugin which can interpret paths and recurse through the DOM.

**Comment by cboulanger** [comment]
I think what they were talking about was to use autogenerated object ids (based on the class name, some increment, and so on.) I also think that could be useful but I would argue for not making this a basic feature of the mechanism because my sense was that there was no consensus about the syntax of how this should be implemented. Rather, we should have implementations based on your solution, and then we can pick an choose which one proves to be most useful.

**Comment by johnspackman** [comment]
Oh, so the `"qx.core.window.Window[2]/qx.ui.form.TextField[3]"` form?  That sounds more like an expression language, but also sounds very fragile (eg minor visual/layout changes break unit tests)

Perhaps we could do something like, first register the window globally (eg as "userProfileEditor") so it can be found, and then assign IDs to each created widget (eg "edtUsername" TextField).  In the window's `.add()` method we could detect that the widget has an ID but does not have an owner and so then hook it up to the window.  This would mean that IDs could be easily retrofitted into an application (which has to be done anyway, and as it's just one more {String} property to set the work would be as simple as it could be), and provided that owned widget is directly added to the window everything will work as expected.

There would be a final piece of the puzzle where fields were added to irrelevant intermediate widgets (like a groupbox) that when it comes to IDs and paths you actually want to ignore completely; so to do this, we'd have a `registerObject(obj...)` API method that would be able to manually attach the widgets directly to the window.

**Comment by cboulanger** [comment]
Automatic Id assignment could also draw form the text of titles, labels etc., but as I said, I would prefer to have this as a separate solution that can be added in on demand.

**Comment by johnspackman** [comment]
Interesting - I don't use forms, but is it typical to associate a label with a form field?  i.e. in the way that a form field can discover it's own label?

**Comment by cboulanger** [comment]
You're right - I haven't worked on my forms for a while. Contrary to what I thought, there is no way to associate labels and forms as there is in HTML with the ["label for" mechanism](https://www.w3schools.com/tags/att_label_for.asp). Maybe there should be.

**Comment by level420** [comment]
If you are using qx.ui.form.Form.add you are adding form fields including their label string values and a field name. The form renderer then creates label widget instances and associates them with the form field by e.g. binding their enabled property so that if you disable the field it also sets the label disabled. I’m using forms a lot.

**Comment by cboulanger** [comment]
[qx.ui.form.CheckBox](http://www.qooxdoo.org/current/api/#qx.ui.form.CheckBox) has a label (as a subclass of [qx.ui.basic.Atom](http://www.qooxdoo.org/current/api/#qx.ui.basic.Atom)), of course, but few others form widgets. Maybe this could be implemented in [qx.ui.core.MForm](http://www.qooxdoo.org/current/api/#qx.ui.form.MForm).

**Comment by cboulanger** [comment]
@level420 of course! the form has the information on fieldnames and labels. Our mechanism could tap into this!

**Comment by level420** [comment]
In fact I’m using a derived form class where I’ve extended the interface to be able to retreive fields by their name. If I’ve added a field ‘username’, I’m able to get itbfrom the form instance via `fom.getItem(‘username’)`.

**Comment by cboulanger** [comment]
Qooxdoo provides a way to bind model properties to a form (see [qx.data.controller.Form](http://www.qooxdoo.org/current/api/#qx.data.controller.Form) - so that the property names are the field names. But it is difficult to auto-discover this because the field widgets no nothing about this controller, and the controller is not in some parent-child relationship to them.

**Comment by johnspackman** [comment]
It would be better to find a way to handle auto ids via Form rather than trying to hook into 'add' and 'remove' because that would impact a lot of widgets and require some careful coding to prevent conflicts with how ids are handled normally.

**Comment by derrell** [comment]
I think I may not have been clear. My goal is that when something goes wrong in my app, to be able to look at the DOM and quickly and easily ascertain how (where from) that DOM element was created. Currently, I have to intuit even what type of widget it is from the `class` value, and then go searching through code to figure out what the possibly many widgets are that use that `class`. I don't (for this purpose) case about a registry at all. All I care about is that there is an ID that is meaningful, preferably on its own, but if necessary (because of child controls or widgets created by widgets), by following its parent path to ascertain which widget _my program instantiated_ that's having the problem.

I don't mind if I have to register the code that generates the IDs, as long as they are created automatically and are there when I discover I need them. That code that I register would presumably be some function that's passed lots of info about whose ID is being generated, so that I can build an appropriate ID.

**Comment by petitben** [comment]
Very interesting topic. I looked for this kind of feature for many years :-)

I know Extjs provides this kind of mechanism using Ext.ComponentQuery. Actually it mimics the functionalities from the standard querySelector. It might worth a look at it for inspiration purpose.

Anyway I wonder if base the work on qx.core.ObjectRegistry would do the trick (as every object can be identified using toHashCode() and toString()).

Voilà, just my 2 cents :-)

**Comment by johnspackman** [comment]
I've never used ExtJS but a quick google and the query mechanism looks interesting; this ID mechanism would be a foundation piece for such a query language, but we should probably consider that as a separate PR

The disadvantage of ObjectRegistry is that having a central list of all objects makes memory management a nightmare, especially for incidental objects - there's a writeup [here](https://github.com/qooxdoo/qooxdoo/issues/215) on the changes we made to remove the reliance on the registry.  The hashcodes are also opaque and not guaranteed repeatable, whereas the ID mechanism here is designed to be human readable (all objects continue to have a hash code though, it's just that they are not registered anywhere by default)

This will be coming up as a PR "real soon now" :)

**Comment by cajus** [comment]
Hmm. I don't understand how your example works. Mustn't it at least call the base `_createObjectImpl` to make the objects beeing registered?

More thoughts: That mechanism isn't valid if you're using *_createChildControlImpl* most of the time to control widget styling. Setting up a form is just a small part of an application, and we even use *_createChildControlImpl* for that here, because it also handles destruction for us. The *_createObjectImpl* code keeps references to the objects that are not removed in the destructor. I don't know whether the automatic memory management will help there, but we had massive problems with non disposed widgets, controllers, etc. in the past that have not been disposed in the destructor. Partly even the destruction order is important if you don't want to run into crashes (mostly list/controller).

Why not just using the *_createChildControlImpl* mechanism for that, too? It's feels much more natural if you've large widget structures IMHO.

**Comment by johnspackman** [comment]
> Hmm. I don't understand how your example works. Mustn't it at least call the base _createObjectImpl to make the objects beeing registered?

It should, you're right - but not to make the objects become registered (calling base would only be necessary if the base class might create widgets on demand also).  When you need to access an object, you call `getObject("edtPostcode")` and then will only call `_createObjectImpl` the first time, to create the object - after that it is cached.  This is the same pattern as `getChildControl`/`_createChildControlImpl`

> More thoughts: That mechanism isn't valid if you're using _createChildControlImpl most of the time to control widget styling. Setting up a form is just a small part of an application, and we even use _createChildControlImpl for that here
Fair point, my mixin here will defer to `getChildControl` if `_createObjectImpl` returned undefined.

> Why not just using the _createChildControlImpl mechanism for that, too? It's feels much more natural if you've large widget structures IMHO.

That was my first attempt years ago but found that there were conflicts with getting the appearances to work, and it might require that everything returned is a `LayoutItem` or `Widget`.  It's a long time ago now and maybe things have moved on, but at the time it caused something to break so I am surprised that works for you.

> The _createObjectImpl code keeps references to the objects that are not removed in the destructor
Yes, that should be fixed too, if `dispose()` is called it should dispose all the owned widgets

If `_createChildControlImpl` can be used for this then I'm less convinced why we want this mechanism because it would mean that there is already a solution for IDs.  The missing piece would be a starting point to resolve absolute ID paths from.

I'll experiment with my mixin to merge it with `_createChildControlImpl` and see if I can reproduce the problems I found before.

**Comment by cboulanger** [comment]
Another use case that just came up in my own code: 

The backend needs to reference an object in the qooxdoo application which cannot be known in advance, but is determined dynamically at runtime. At the moment, I solve this by `eval`ing code, which is obviously bad. A deterministic object id would allow to safely reference this object and preserve isolation of application components.

@johnspackman, did you have a chance to experiment with child controls? A difference between parent/child components and child controls would be that the child control know nothing about their parent control (or do they?), whereas child components should have a way to look up their parent.

**Comment by johnspackman** [comment]
No not yet - it's still on my list.  I'm keen to do it because I think I have such a good set of test cases, it's getting closer to the top of the todo list but not as fast as I'd like :(

My proposed implementation does have a parent, it's called "owner" so you can navigate up and down

**Comment by johnspackman** [comment]
Earlier this year I finally tried @cajus' suggestion to use `getChildControl`/`_createChildControlImpl` mechanism to access widgets and create them on demand - I can't reproduce those early problems I had (and it was a long time ago).

While it is possible to use the childControl mechanism, I don’t think that’s the right solution mostly because of a “separation of concerns” argument.  Primarily, I see the object ID mechanism as a means to navigate *application-level* objects (aka business objects, etc) from outside the application or component - this makes the IDs a form of published API that can expose a specific, designed structure, and as an “API” it needs to be clear and consistent.  An API like this can be used by Selenium-type tools, or as part of the application’s internal structure.

The `childControl` mechanism is tied into appearances and is a much lower level - when looking at a `qx.ui.form.SelectBox`, it’s comprised of a list, a button, a popup window, etc - while those objects are published within the context of *just* a `SelectBox`, they make little sense when it comes to application testing or finding application objects.  ChildControls even have their own relationship with documentation (eg `@childControl` jsdoc tag), and just the term “control” implies some kind of visual aspect.

@cajus mentioned that `childControl`’s have their destructor called automatically when the layout parent is destroyed - I can see the advantage of that, especially before Automatic Memory Management, but it would be problematic in some circumstances because it means that your object's lifetime is tied to the lifetime of the UI widget that displays it.  It also means that objects cannot appear in your object graph more than once.

IMHO `childControl` and `objectId` are for distinctly different purposes (ie one is presentation layer and the other is data/business object layer) and should be kept separate.  And if you do need to find a child control, there’s no reason why you can't find the object via an objectId and then look inside it’s childControl mechanism.

There is a potential cost to a parallel objectId mechanism, but it’s very tiny - at most one additional map - but bear in mind that objects are not stored twice, just stored in one of two places (according to their purpose), so the cost is minimal.  The PR includes the ability to turn off the objectId mechanism altogether via a `qx.core.Environment` setting.

So in conclusion, IMHO we should have an `objectId` mechanism which is separate from `childControl` and I’ve continued this PR on that basis (BTW I am slightly biased 🙂, in that I have used an objectId mechanism for many years, and run it strictly in parallel to the childControl mechanism)

I’ve updated the PR with a new commit that adds globally unique IDs to the DOM elements, and adds a UI test that demonstrates the mechanism.  This means that the DOM elements have a unique ID which the developer will (hopefully) recognise intuitively (eg “application/mainForm/profileEditor/edtName”) and which crucially can also can be located by Selenium.

**Comment by cboulanger** [comment]
As noted somewhere above, I fully agree that object ids and child controls have different purposes and should be implemented differently. I think your implementation is sound and minimal/unopinionated enough to find consensus. On this basis, I could contribute a contrib that provides support for Puppeteer UI tests.

**Comment by cboulanger** [comment]
Implemented by #9594.
